File: src/main/java/com/yupi/airouter/controller/BalanceController.java
package com.yupi.airouter.controller;

import com.mybatisflex.core.paginate.Page;
import com.yupi.airouter.annotation.AuthCheck;
import com.yupi.airouter.common.BaseResponse;
import com.yupi.airouter.common.ResultUtils;
import com.yupi.airouter.constant.UserConstant;
import com.yupi.airouter.model.entity.BillingRecord;
import com.yupi.airouter.model.entity.User;
import com.yupi.airouter.service.BalanceService;
import com.yupi.airouter.service.BillingRecordService;
import com.yupi.airouter.service.UserService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.annotation.Resource;
import jakarta.servlet.http.HttpServletRequest;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.*;

import java.io.Serializable;
import java.math.BigDecimal;

/**
 * 余额管理控制器
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
@RestController
@RequestMapping("/balance")
@Tag(name = "余额管理")
@Slf4j
public class BalanceController {

    @Resource
    private BalanceService balanceService;

    @Resource
    private BillingRecordService billingRecordService;

    @Resource
    private UserService userService;

    /**
     * 获取我的余额信息
     */
    @GetMapping("/my")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "获取我的余额信息")
    public BaseResponse<BalanceVO> getMyBalance(HttpServletRequest request) {
        User loginUser = userService.getLoginUser(request);
        BigDecimal balance = balanceService.getUserBalance(loginUser.getId());
        BigDecimal totalSpending = billingRecordService.getUserTotalSpending(loginUser.getId());
        BigDecimal totalRecharge = billingRecordService.getUserTotalRecharge(loginUser.getId());

        BalanceVO balanceVO = new BalanceVO();
        balanceVO.setBalance(balance);
        balanceVO.setTotalSpending(totalSpending);
        balanceVO.setTotalRecharge(totalRecharge);

        return ResultUtils.success(balanceVO);
    }

    /**
     * 获取我的消费账单（分页）
     */
    @GetMapping("/billing/my")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "获取我的消费账单")
    public BaseResponse<Page<BillingRecord>> getMyBillingRecords(
            @RequestParam(defaultValue = "1") int pageNum,
            @RequestParam(defaultValue = "10") int pageSize,
            HttpServletRequest request) {
        User loginUser = userService.getLoginUser(request);
        Page<BillingRecord> page = billingRecordService.listUserBillingRecords(
                loginUser.getId(), pageNum, pageSize);
        return ResultUtils.success(page);
    }

    /**
     * 余额信息VO
     */
    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class BalanceVO implements Serializable {
        private BigDecimal balance; // 当前余额
        private BigDecimal totalSpending; // 总消费
        private BigDecimal totalRecharge; // 总充值
    }
}

File: go-backend/internal/controller/balance_controller.go
package controller

import (
	"log"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/yupi/airouter/go-backend/internal/common"
	"github.com/yupi/airouter/go-backend/internal/errno"
	"github.com/yupi/airouter/go-backend/internal/model/vo"
	"github.com/yupi/airouter/go-backend/internal/service"
)

type BalanceController struct {
	balanceService       *service.BalanceService
	billingRecordService *service.BillingRecordService
	userService          *service.UserService
}

func NewBalanceController(
	balanceService *service.BalanceService,
	billingRecordService *service.BillingRecordService,
	userService *service.UserService,
) *BalanceController {
	return &BalanceController{
		balanceService:       balanceService,
		billingRecordService: billingRecordService,
		userService:          userService,
	}
}

func (b *BalanceController) GetMyBalance(c *gin.Context) {
	loginUser, err := b.userService.GetLoginUser(c)
	if err != nil {
		if bizErr, ok := errno.AsBusinessError(err); ok {
			common.Error(c, bizErr.Code, bizErr.Message)
			return
		}
		common.Error(c, errno.SystemError.Code, errno.SystemError.Message)
		return
	}
	balance, err := b.balanceService.GetUserBalance(loginUser.ID)
	if err != nil {
		b.handleError(c, err)
		return
	}
	totalSpending, err := b.billingRecordService.GetUserTotalSpending(loginUser.ID)
	if err != nil {
		b.handleError(c, err)
		return
	}
	totalRecharge, err := b.billingRecordService.GetUserTotalRecharge(loginUser.ID)
	if err != nil {
		b.handleError(c, err)
		return
	}
	common.Success(c, vo.BalanceVO{
		Balance:       balance,
		TotalSpending: totalSpending,
		TotalRecharge: totalRecharge,
	})
}

func (b *BalanceController) GetMyBillingRecords(c *gin.Context) {
	loginUser, err := b.userService.GetLoginUser(c)
	if err != nil {
		b.handleError(c, err)
		return
	}
	pageNum := int64(1)
	pageSize := int64(10)
	if raw := c.Query("pageNum"); raw != "" {
		if parsed, parseErr := strconv.ParseInt(raw, 10, 64); parseErr == nil && parsed > 0 {
			pageNum = parsed
		}
	}
	if raw := c.Query("pageSize"); raw != "" {
		if parsed, parseErr := strconv.ParseInt(raw, 10, 64); parseErr == nil && parsed > 0 {
			pageSize = parsed
		}
	}
	page, err := b.billingRecordService.ListUserBillingRecords(loginUser.ID, pageNum, pageSize)
	if err != nil {
		b.handleError(c, err)
		return
	}
	common.Success(c, page)
}

func (b *BalanceController) handleError(c *gin.Context, err error) {
	if bizErr, ok := errno.AsBusinessError(err); ok {
		common.Error(c, bizErr.Code, bizErr.Message)
		return
	}
	log.Printf("balance controller error: %v", err)
	common.Error(c, errno.SystemError.Code, errno.SystemError.Message)
}
